File: CBORExtendedFloat.cs
/*
Written in 2014 by Peter O.
Any copyright is dedicated to the Public Domain.
http://creativecommons.org/publicdomain/zero/1.0/
If you like this, you should donate to Peter O.
at: http://peteroupc.github.io/CBOR/
 */
using System;

namespace PeterO
{
  internal class CBORExtendedFloat : ICBORNumber
  {
    
    public bool IsPositiveInfinity(object obj)
    {
      ExtendedFloat ef=(ExtendedFloat)obj;
      return ef.IsPositiveInfinity();
    }
    
    public bool IsInfinity(object obj)
    {
      ExtendedFloat ef=(ExtendedFloat)obj;
      return ef.IsInfinity();
    }
    
    public bool IsNegativeInfinity(object obj)
    {
      ExtendedFloat ef=(ExtendedFloat)obj;
      return ef.IsNegativeInfinity();
    }
    
    public bool IsNaN(object obj)
    {
      ExtendedFloat ef=(ExtendedFloat)obj;
      return ef.IsNaN();
    }
    
    public double AsDouble(object obj)
    {
      ExtendedFloat ef=(ExtendedFloat)obj;
      return ef.ToDouble();
    }
    
    public ExtendedDecimal AsExtendedDecimal(object obj)
    {
      ExtendedFloat ef=(ExtendedFloat)obj;
      return ef.ToExtendedDecimal();
    }
    
    public ExtendedFloat AsExtendedFloat(object obj)
    {
      ExtendedFloat ef=(ExtendedFloat)obj;
      return ef;
    }
    
    public float AsSingle(object obj)
    {
      ExtendedFloat ef=(ExtendedFloat)obj;
      return ef.ToSingle();
    }
    
    public BigInteger AsBigInteger(object obj)
    {
      ExtendedFloat ef=(ExtendedFloat)obj;
      return ef.ToBigInteger();
    }
    
    public long AsInt64(object obj)
    {
      ExtendedFloat ef=(ExtendedFloat)obj;
      if (ef.IsFinite) {
        BigInteger bi = ef.ToBigInteger();
        if (bi.CompareTo(CBORObject.Int64MaxValue) <= 0 &&
            bi.CompareTo(CBORObject.Int64MinValue) >= 0) {
          return (long)bi;
        }
      }
      throw new OverflowException("This object's value is out of range");
    }
    
    public bool CanFitInSingle(object obj)
    {
      ExtendedFloat ef=(ExtendedFloat)obj; throw new NotImplementedException();
    }
    
    public bool CanFitInDouble(object obj)
    {
      ExtendedFloat ef=(ExtendedFloat)obj; throw new NotImplementedException();
    }
    
    public bool CanFitInInt32(object obj)
    {
      return IsIntegral(obj) && CanTruncatedIntFitInInt32(obj);
    }
    
    public bool CanFitInInt64(object obj)
    {
      return IsIntegral(obj) && CanTruncatedIntFitInInt64(obj);
    }
    
    public bool CanTruncatedIntFitInInt64(object obj)
    {
      ExtendedFloat ef=(ExtendedFloat)obj; throw new NotImplementedException();
    }
    
    public bool CanTruncatedIntFitInInt32(object obj)
    {
      ExtendedFloat ef=(ExtendedFloat)obj; throw new NotImplementedException();
    }
    
    public int AsInt32(object obj)
    {
      ExtendedFloat ef=(ExtendedFloat)obj; throw new NotImplementedException();
    }
    
    public bool IsZero(object obj)
    {
      ExtendedFloat ef=(ExtendedFloat)obj; throw new NotImplementedException();
    }
    
    public int Sign(object obj)
    {
      ExtendedFloat ef=(ExtendedFloat)obj; throw new NotImplementedException();
    }
    
    public bool IsIntegral(object obj)
    {
      ExtendedFloat ef=(ExtendedFloat)obj;
      if (!ef.IsFinite) {
        return false;
      }
      if (ef.Exponent.Sign >= 0) {
        return true;
      }
      ExtendedFloat ef2 = ExtendedFloat.FromBigInteger(ef.ToBigInteger());
      return ef2.CompareTo(ef) == 0;
    }
    
  public bool CanFitInTypeZeroOrOne(object obj)
  {
    throw new NotImplementedException();
  }
    
  public int AsInt32(object obj, int minValue, int maxValue)
  {
    throw new NotImplementedException();
  }
  }
}

File: src/main/java/com/upokecenter/cbor/CBORExtendedFloat.java
package com.upokecenter.cbor;
/*
Written in 2014 by Peter O.
Any copyright is dedicated to the Public Domain.
http://creativecommons.org/publicdomain/zero/1.0/
If you like this, you should donate to Peter O.
at: http://upokecenter.com/d/
 */

import com.upokecenter.util.*;

  class CBORExtendedFloat implements ICBORNumber {

    public boolean IsPositiveInfinity(Object obj) {
      ExtendedFloat ef = (ExtendedFloat)obj;
      return ef.IsPositiveInfinity();
    }

    public boolean IsInfinity(Object obj) {
      ExtendedFloat ef = (ExtendedFloat)obj;
      return ef.IsInfinity();
    }

    public boolean IsNegativeInfinity(Object obj) {
      ExtendedFloat ef = (ExtendedFloat)obj;
      return ef.IsNegativeInfinity();
    }

    public boolean IsNaN(Object obj) {
      ExtendedFloat ef = (ExtendedFloat)obj;
      return ef.IsNaN();
    }

    public double AsDouble(Object obj) {
      ExtendedFloat ef = (ExtendedFloat)obj;
      return ef.ToDouble();
    }

    public ExtendedDecimal AsExtendedDecimal(Object obj) {
      ExtendedFloat ef = (ExtendedFloat)obj;
      return ef.ToExtendedDecimal();
    }

    public ExtendedFloat AsExtendedFloat(Object obj) {
      ExtendedFloat ef = (ExtendedFloat)obj;
      return ef;
    }

    public float AsSingle(Object obj) {
      ExtendedFloat ef = (ExtendedFloat)obj;
      return ef.ToSingle();
    }

    public BigInteger AsBigInteger(Object obj) {
      ExtendedFloat ef = (ExtendedFloat)obj;
      return ef.ToBigInteger();
    }

    public long AsInt64(Object obj) {
      ExtendedFloat ef = (ExtendedFloat)obj;
      if (this.CanTruncatedIntFitInInt64(obj)) {
        BigInteger bi = ef.ToBigInteger();
        return bi.longValue();
      }
      throw new ArithmeticException("This Object's value is out of range");
    }

    public boolean CanFitInSingle(Object obj) {
      ExtendedFloat ef = (ExtendedFloat)obj;
      return (!ef.isFinite()) ||
      (ef.compareTo(ExtendedFloat.FromSingle(ef.ToSingle())) == 0);
    }

    public boolean CanFitInDouble(Object obj) {
      ExtendedFloat ef = (ExtendedFloat)obj;
      return (!ef.isFinite()) ||
      (ef.compareTo(ExtendedFloat.FromDouble(ef.ToDouble())) == 0);
    }

    public boolean CanFitInInt32(Object obj) {
      return this.IsIntegral(obj) && this.CanTruncatedIntFitInInt32(obj);
    }

    public boolean CanFitInInt64(Object obj) {
      return this.IsIntegral(obj) && this.CanTruncatedIntFitInInt64(obj);
    }

    public boolean CanTruncatedIntFitInInt64(Object obj) {
      ExtendedFloat ef = (ExtendedFloat)obj;
      if (!ef.isFinite()) {
        return false;
      }
      if (ef.signum() == 0) {
        return true;
      }
      if (ef.getExponent().compareTo(BigInteger.valueOf(65)) >= 0) {
        return false;
      }
      BigInteger bi = ef.ToBigInteger();
      return bi.bitLength() <= 63;
    }

    public boolean CanTruncatedIntFitInInt32(Object obj) {
      ExtendedFloat ef = (ExtendedFloat)obj;
      if (!ef.isFinite()) {
        return false;
      }
      if (ef.signum() == 0) {
        return true;
      }
      if (ef.getExponent().compareTo(BigInteger.valueOf(33)) >= 0) {
        return false;
      }
      BigInteger bi = ef.ToBigInteger();
      return bi.canFitInInt();
    }

    public boolean IsZero(Object obj) {
      ExtendedFloat ef = (ExtendedFloat)obj;
      return ef.signum() == 0;
    }

    public int Sign(Object obj) {
      ExtendedFloat ef = (ExtendedFloat)obj;
      return ef.IsNaN() ? 2 : ef.signum();
    }

    public boolean IsIntegral(Object obj) {
      ExtendedFloat ef = (ExtendedFloat)obj;
      if (!ef.isFinite()) {
        return false;
      }
      if (ef.getExponent().signum() >= 0) {
        return true;
      }
      ExtendedFloat ef2 = ExtendedFloat.FromBigInteger(ef.ToBigInteger());
      return ef2.compareTo(ef) == 0;
    }

    public int AsInt32(Object obj, int minValue, int maxValue) {
      ExtendedFloat ef = (ExtendedFloat)obj;
      if (this.CanTruncatedIntFitInInt32(obj)) {
        BigInteger bi = ef.ToBigInteger();
        int ret = bi.intValue();
        if (ret >= minValue && ret <= maxValue) {
          return ret;
        }
      }
      throw new ArithmeticException("This Object's value is out of range");
    }

    public Object Negate(Object obj) {
      ExtendedFloat ed = (ExtendedFloat)obj;
      return ed.Negate();
    }

    public Object Abs(Object obj) {
      ExtendedFloat ed = (ExtendedFloat)obj;
      return ed.Abs();
    }

    public ExtendedRational AsExtendedRational(Object obj) {
      return ExtendedRational.FromExtendedFloat((ExtendedFloat)obj);
    }
  }
